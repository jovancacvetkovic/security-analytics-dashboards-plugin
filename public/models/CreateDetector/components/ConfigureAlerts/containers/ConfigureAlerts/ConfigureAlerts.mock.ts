/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { mockDetector } from '../../../../../Interfaces.mock';
import { mockRuleOptions } from '../../components/AlertCondition/AlertConditionPanel.mock';
import { NotificationsService } from '../../../../../../services';

export const mockNotificationsService: NotificationsService = {
  getChannels: () => {
    return {
      ok: true,
      response: {
        channel_list: [],
      },
    };
  },
};

export default {
  detector: mockDetector,
  isEdit: false,
  rulesOptions: [mockRuleOptions],
  changeDetector: jest.fn(),
  updateDataValidState: jest.fn(),
  notificationsService: mockNotificationsService,
  hasNotificationPlugin: false,
};
