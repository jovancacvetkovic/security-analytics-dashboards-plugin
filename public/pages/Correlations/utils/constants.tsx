/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { CorrelationGraphData } from '../../../../types';
import { ruleSeverity, ruleTypes } from '../../Rules/utils/constants';
import { FilterItem } from '../components/FilterGroup';
import { EuiIcon } from '@elastic/eui';

export const graphRenderOptions = {
  nodes: {
    shape: 'circle',
  },
  edges: {
    arrows: {
      to: {
        enabled: false,
      },
    },
    color: '#000000',
    chosen: true,
  },
  layout: {
    hierarchical: false,
    randomSeed: 2222,
    improvedLayout: false,
  },
  autoResize: true,
  height: '800px',
  width: '100%',
  physics: {
    stabilization: {
      fit: true,
      iterations: 1000,
    },
  },
  interaction: {
    zoomView: true,
    zoomSpeed: 0.2,
    dragView: true,
    dragNodes: false,
    multiselect: true,
    tooltipDelay: 50,
    hover: true,
  },
};

export const defaultLogTypeFilterItemOptions: FilterItem[] = Object.values(ruleTypes).map(
  (type) => {
    return {
      name: `${type.abbr}: ${type.label}`,
      id: type.value,
      checked: 'on',
    };
  }
);

export const defaultSeverityFilterItemOptions: FilterItem[] = Object.values(ruleSeverity).map(
  (sev) => {
    return {
      name: (
        <p>
          <EuiIcon type={'dot'} color={sev.color} /> {`${sev.priority} ${sev.name}`}
        </p>
      ),
      id: sev.value,
      checked: 'on',
    };
  }
);

export const emptyGraphData: CorrelationGraphData = {
  graph: {
    nodes: [],
    edges: [],
  },
  events: {},
};

export const getAbbrFromLogType = (logType: string) => {
  return ruleTypes.find((ruleType) => ruleType.value === logType)?.abbr || '-';
};

export const getLabelFromLogType = (logType: string) => {
  return ruleTypes.find((ruleType) => ruleType.value === logType)?.label || '-';
};

export const getSeverityLabel = (sev: string) => {
  return ruleSeverity.find((severity) => severity.value === sev)?.name || '';
};

export const getSeverityColor = (sev: string) => {
  return ruleSeverity.find((severity) => severity.value === sev)?.color || 'black';
};
