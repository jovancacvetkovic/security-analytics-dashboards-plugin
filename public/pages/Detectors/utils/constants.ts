/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const MAX_DETECTOR_INPUTS = 1;
export const MIN_NUM_DATA_SOURCES = 1;

export const STATUS_OPTIONS = [
  { value: 'test_status1', text: 'Test status #1' },
  { value: 'test_status2', text: 'Test status #2' },
];

export const EMPTY_DEFAULT_PERIOD_SCHEDULE = {
  period: {
    interval: 1,
    unit: 'm',
  },
};

export const EMPTY_DEFAULT_DETECTOR_INPUT = {
  input: {
    description: '',
    indices: [],
    rules: [],
  },
};

export const DETECTOR_TYPES = {
  NETWORK: { id: 'network', label: 'Network events', abbr: 'NTW' },
  DNS: { id: 'dns', label: 'DNS logs', abbr: 'DNS' },
  APACHE_ACCESS: { id: 'apache_access', label: 'Apache access logs', abbr: 'APC' },
  WINDOWS: { id: 'windows', label: 'Windows logs', abbr: 'WIN' },
  AD_LDAP: { id: 'ad_ldap', label: 'AD/LDAP logs', abbr: 'AD' },
  SYSTEM: { id: 'linux', label: 'System logs', abbr: 'LNX' },
  CLOUD_TRAIL: { id: 'cloudtrail', label: 'Cloud Trail logs', abbr: 'CLT' },
  S3: { id: 's3', label: 'S3 access logs', abbr: 'S3' },
  GWORKSPACE: { id: 'gworkspace', label: 'Google Workspace logs', abbr: 'GGL' },
  GITHUB: { id: 'github', label: 'Github actions', abbr: 'GHB' },
  M365: { id: 'm365', label: 'Microsoft 365 logs', abbr: 'MSO' },
  OKTA: { id: 'okta', label: 'Okta events', abbr: 'OKT' },
  AZURE: { id: 'azure', label: 'Azure logs', abbr: 'AZR' },
};
