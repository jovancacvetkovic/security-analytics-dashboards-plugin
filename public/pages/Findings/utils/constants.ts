/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export enum FindingFlyoutTabId {
  DETAILS = 'Details',
  CORRELATIONS = 'Correlations',
}

export const FindingFlyoutTabs = [
  { id: FindingFlyoutTabId.DETAILS, name: 'Details' },
  { id: FindingFlyoutTabId.CORRELATIONS, name: 'Correlations' },
];
