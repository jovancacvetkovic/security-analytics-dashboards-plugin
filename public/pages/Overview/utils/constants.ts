/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const summaryGroupByOptions = [
  { text: 'All findings', value: 'finding' },
  { text: 'Log type', value: 'logType' },
];
