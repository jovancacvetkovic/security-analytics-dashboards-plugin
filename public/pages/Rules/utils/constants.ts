/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { euiPaletteForStatus } from '@elastic/eui';

export const ruleTypes: { label: string; value: string; abbr: string }[] = [
  { abbr: 'NTW', label: 'Network', value: 'network' },
  { abbr: 'DNS', label: 'DNS', value: 'dns' },
  { abbr: 'APC', label: 'Apache Access', value: 'apache_access' },
  { abbr: 'WIN', label: 'Windows', value: 'windows' },
  { abbr: 'AD', label: 'AD/LDAP', value: 'ad_ldap' },
  { abbr: 'LNX', label: 'Linux', value: 'linux' },
  { abbr: 'CLT', label: 'Cloudtrail', value: 'cloudtrail' },
  { abbr: 'S3', label: 'S3', value: 's3' },
  { abbr: 'GGL', label: 'Google Workspace', value: 'gworkspace' },
  { abbr: 'GHB', label: 'Github actions', value: 'github' },
  { abbr: 'MSO', label: 'Microsoft 365', value: 'm365' },
  { abbr: 'OKT', label: 'Okta', value: 'okta' },
  { abbr: 'AZR', label: 'Azure', value: 'azure' },
];

const paletteColors = euiPaletteForStatus(5);

export const ruleSeverity: { name: string; value: string; priority: string; color: string }[] = [
  { name: 'Critical', value: 'critical', priority: '1', color: paletteColors[4] },
  { name: 'High', value: 'high', priority: '2', color: paletteColors[3] },
  { name: 'Medium', value: 'medium', priority: '3', color: paletteColors[2] },
  { name: 'Low', value: 'low', priority: '4', color: paletteColors[1] },
  { name: 'Informational', value: 'informational', priority: '5', color: paletteColors[0] },
];

export const ruleSource: string[] = ['Sigma', 'Custom'];

export const ruleStatus: string[] = ['experimental', 'test', 'stable'];
