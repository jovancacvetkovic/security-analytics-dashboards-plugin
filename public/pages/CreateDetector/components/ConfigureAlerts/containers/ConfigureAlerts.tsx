/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { Component } from 'react';
import { RouteComponentProps } from 'react-router-dom';
import {
  EuiAccordion,
  EuiButton,
  EuiCallOut,
  EuiFlexGroup,
  EuiFlexItem,
  EuiHorizontalRule,
  EuiPanel,
  EuiSpacer,
  EuiText,
  EuiTitle,
} from '@elastic/eui';
import { MAX_ALERT_CONDITIONS } from '../utils/constants';
import AlertConditionPanel from '../components/AlertCondition';
import { DetectorCreationStep } from '../../../models/types';
import { CreateDetectorRulesOptions } from '../../../../../models/types';
import { NotificationChannelTypeOptions } from '../models/interfaces';
import {
  getEmptyAlertCondition,
  getNotificationChannels,
  parseNotificationChannelsToOptions,
} from '../utils/helpers';
import { NotificationsService } from '../../../../../services';
import { validateName } from '../../../../../utils/validation';
import { CoreServicesContext } from '../../../../../components/core_services';
import { BREADCRUMBS } from '../../../../../utils/constants';
import { Detector } from '../../../../../../types';

interface ConfigureAlertsProps extends RouteComponentProps {
  detector: Detector;
  isEdit: boolean;
  rulesOptions: CreateDetectorRulesOptions;
  changeDetector: (detector: Detector) => void;
  updateDataValidState: (step: DetectorCreationStep, isValid: boolean) => void;
  notificationsService: NotificationsService;
  hasNotificationPlugin: boolean;
  skipAndConfigureHandler: () => void;
}

interface ConfigureAlertsState {
  loading: boolean;
  notificationChannels: NotificationChannelTypeOptions[];
}

export default class ConfigureAlerts extends Component<ConfigureAlertsProps, ConfigureAlertsState> {
  static contextType = CoreServicesContext;

  constructor(props: ConfigureAlertsProps) {
    super(props);
    this.state = {
      loading: false,
      notificationChannels: [],
    };
  }

  updateBreadcrumbs = () => {
    const {
      isEdit,
      detector: { id = '', name },
    } = this.props;

    isEdit &&
      this.context.chrome.setBreadcrumbs([
        BREADCRUMBS.SECURITY_ANALYTICS,
        BREADCRUMBS.DETECTORS,
        BREADCRUMBS.DETECTORS_DETAILS(name, id),
        {
          text: 'Edit alert triggers',
        },
      ]);
  };

  componentDidMount = async () => {
    this.updateBreadcrumbs();
    const {
      detector: { triggers },
    } = this.props;
    this.getNotificationChannels();

    if (triggers.length === 0) {
      this.addCondition();
    }
  };

  componentDidUpdate(
    prevProps: Readonly<ConfigureAlertsProps>,
    prevState: Readonly<ConfigureAlertsState>,
    snapshot?: any
  ) {
    this.updateBreadcrumbs();
  }

  getNotificationChannels = async () => {
    this.setState({ loading: true });
    const channels = await getNotificationChannels(this.props.notificationsService);
    this.setState({ notificationChannels: parseNotificationChannelsToOptions(channels) });
    this.setState({ loading: false });
  };

  addCondition = () => {
    const {
      changeDetector,
      detector,
      detector: { triggers },
    } = this.props;
    triggers.push(getEmptyAlertCondition());
    changeDetector({ ...detector, triggers });
  };

  onAlertTriggerChanged = (newDetector: Detector): void => {
    const isTriggerDataValid =
      !newDetector.triggers.length ||
      newDetector.triggers.every((trigger) => {
        return !!trigger.name && validateName(trigger.name) && trigger.severity;
      });

    this.props.changeDetector(newDetector);
    this.props.updateDataValidState(DetectorCreationStep.CONFIGURE_ALERTS, isTriggerDataValid);
  };

  onDelete = (index: number) => {
    const {
      detector,
      detector: { triggers },
    } = this.props;
    triggers.splice(index, 1);
    this.onAlertTriggerChanged({ ...detector, triggers: triggers });
  };

  render() {
    const {
      isEdit,
      detector: { triggers },
      skipAndConfigureHandler,
    } = this.props;

    let getPageTitle = (): string | JSX.Element => {
      if (isEdit) {
        return <>Alert triggers (${triggers.length})</>;
      }

      return (
        <EuiFlexGroup alignItems={'center'}>
          <EuiFlexItem grow={true}>
            <EuiTitle size={'m'}>
              <h3>Set up alert triggers</h3>
            </EuiTitle>
            <EuiText size="s" color="subdued">
              Get notified when specific rule conditions are found by the detector.
            </EuiText>
          </EuiFlexItem>
          {triggers?.length && (
            <EuiFlexItem grow={false}>
              <EuiButton
                onClick={() => {
                  const { changeDetector, detector } = this.props;
                  changeDetector({ ...detector, triggers: [] });
                  skipAndConfigureHandler();
                }}
              >
                Skip and configure later
              </EuiButton>
            </EuiFlexItem>
          )}
        </EuiFlexGroup>
      );
    };

    const { loading, notificationChannels } = this.state;
    return (
      <div>
        {getPageTitle()}

        <EuiSpacer size={'m'} />

        {triggers.map((alertCondition, index) => (
          <div key={index}>
            {index > 0 && <EuiSpacer size={'l'} />}
            <EuiPanel>
              <EuiAccordion
                id={`alert-condition-${index}`}
                buttonContent={
                  <EuiTitle>
                    <h4>{isEdit ? alertCondition.name : 'Alert trigger'}</h4>
                  </EuiTitle>
                }
                paddingSize={'none'}
                initialIsOpen={true}
                extraAction={
                  <EuiButton color="danger" onClick={() => this.onDelete(index)}>
                    Remove
                  </EuiButton>
                }
              >
                <EuiHorizontalRule margin={'xs'} />
                <EuiSpacer size={'m'} />
                <AlertConditionPanel
                  {...this.props}
                  alertCondition={alertCondition}
                  allNotificationChannels={notificationChannels}
                  indexNum={index}
                  loadingNotifications={loading}
                  onAlertTriggerChanged={this.onAlertTriggerChanged}
                  refreshNotificationChannels={this.getNotificationChannels}
                  hasNotificationPlugin={this.props.hasNotificationPlugin}
                />
              </EuiAccordion>
            </EuiPanel>
          </div>
        ))}
        {!triggers?.length && (
          <EuiCallOut
            color={'primary'}
            iconType={'iInCircle'}
            title={
              <>
                <p>
                  We recommend creating alert triggers to get notified when specific conditions are
                  found by the detector.
                </p>
                <p>You can also configure alert triggers after the detector is created.</p>
              </>
            }
          />
        )}

        <EuiSpacer size={'m'} />

        <EuiButton disabled={triggers.length >= MAX_ALERT_CONDITIONS} onClick={this.addCondition}>
          {triggers.length > 0 ? 'Add another alert trigger' : 'Add alert triggers'}
        </EuiButton>
      </div>
    );
  }
}
