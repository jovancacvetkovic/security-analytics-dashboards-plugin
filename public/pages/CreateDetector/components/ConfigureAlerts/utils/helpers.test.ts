import {
  createSelectedOptions,
  getEmptyAlertCondition,
  getNotificationChannels,
  parseAlertSeverityToOption,
  parseNotificationChannelsToOptions,
} from './helpers';
import { expect } from '@jest/globals';
import { mockNotificationsService } from '../../../../../models/CreateDetector/components/ConfigureAlerts/containers/ConfigureAlerts/ConfigureAlerts.mock';
import { CHANNEL_TYPES } from './constants';

describe('helper utilities spec', () => {
  describe('tests parseAlertSeverityToOption function', () => {
    it(' - should match the snapshot', () => {
      expect(parseAlertSeverityToOption('1')).toMatchSnapshot();
    });
  });

  describe('tests createSelectedOptions function', () => {
    it(' - should match the snapshot', () => {
      expect(createSelectedOptions(['1 Highest'])).toMatchSnapshot();
    });
  });

  describe('tests getNotificationChannels function', () => {
    it(' - should match the snapshot', () => {
      expect(getNotificationChannels(mockNotificationsService)).toMatchSnapshot();
    });
  });

  describe('tests getEmptyAlertCondition function', () => {
    it(' - should match the snapshot', () => {
      expect(getEmptyAlertCondition()).toMatchSnapshot();
    });
  });

  describe('tests parseNotificationChannelsToOptions function', () => {
    it(' - should match the snapshot', () => {
      let notificationChannels = [
        {
          config_id: 'mockConfigId',
          name: 'mockName',
          description: 'mockDescription',
          config_type: 'mockConfigType',
          is_enabled: false,
        },
      ];
      expect(
        parseNotificationChannelsToOptions(notificationChannels, CHANNEL_TYPES)
      ).toMatchSnapshot();
    });
  });
});
