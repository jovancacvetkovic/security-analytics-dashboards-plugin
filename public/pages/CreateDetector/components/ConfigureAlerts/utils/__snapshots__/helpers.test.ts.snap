// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helper utilities spec tests createSelectedOptions function  - should match the snapshot 1`] = `
Array [
  Object {
    "id": "1 Highest",
    "label": "1 Highest",
  },
]
`;

exports[`helper utilities spec tests getEmptyAlertCondition function  - should match the snapshot 1`] = `
Object {
  "actions": Array [
    Object {
      "destination_id": "",
      "id": "",
      "message_template": Object {
        "lang": "mustache",
        "source": "",
      },
      "name": "",
      "subject_template": Object {
        "lang": "mustache",
        "source": "",
      },
      "throttle": Object {
        "unit": "MINUTES",
        "value": 10,
      },
      "throttle_enabled": false,
    },
  ],
  "ids": Array [],
  "name": "",
  "sev_levels": Array [],
  "severity": "1",
  "tags": Array [],
  "types": Array [],
}
`;

exports[`helper utilities spec tests getNotificationChannels function  - should match the snapshot 1`] = `Promise {}`;

exports[`helper utilities spec tests parseAlertSeverityToOption function  - should match the snapshot 1`] = `
Object {
  "id": "1",
  "label": "1 (Highest)",
  "text": "1 (Highest)",
  "value": "1",
}
`;

exports[`helper utilities spec tests parseNotificationChannelsToOptions function  - should match the snapshot 1`] = `
Array [
  Object {
    "label": "slack",
    "options": Array [],
  },
  Object {
    "label": "email",
    "options": Array [],
  },
  Object {
    "label": "chime",
    "options": Array [],
  },
  Object {
    "label": "webhook",
    "options": Array [],
  },
  Object {
    "label": "ses",
    "options": Array [],
  },
  Object {
    "label": "sns",
    "options": Array [],
  },
]
`;
